Unique list

dhruvsomani opened the pull request:
Added the `--list-unique` command line argument which enables to only see the unique errors in a report only manner. Thus, recurring not matching tokens will be lumped together as one. Also, shows a list of line numbers beside the token where the token was used.

The list_unique parameter was added to spell_check(), spell_check_file and spell_check_token() functions.

Closes https://github.com/myint/scspell/issues/24

jayvdb commented on scspell/__init__.py:
```diff
@@ -965,5 +995,6 @@ def main():
-        return 0 if okay else 1
+                           args.test_input,
+                           args.list_unique)
+        return 0 if okay else 1
```
you remove the EOL here.

maybe you need to use a different editor to re-add this.
`vi` definitely lets you control this.

there is a vim for windows, if you are so afflicted.

then use `git commit --amend -a` to re-create this commit , and the git push , with `--force` to overwrite this commit.

jayvdb commented on scspell/__init__.py:
```diff
@@ -572,13 +581,30 @@ def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+
+        print("Set of mispelled tokens:")
+        for row in errors:
+            print(filename + ":", row[1], ":", row[0].lstrip())
```
insead of using `,` , use % notation, like you see in the `print` calls in `report_failed_check`

jayvdb commented on scspell/__init__.py:
```diff
@@ -572,13 +581,30 @@ def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+        errors = list(errors.items())
+        errors.sort(key=lambda x: x[1])
+
+        print("Set of mispelled tokens:")
```
single quotation marks please

jayvdb commented on scspell/__init__.py:
```diff
@@ -556,13 +581,32 @@ def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+
+        print('Set of mispelled tokens:')
+        for row in errors:
+            print("%s:[%s]: %s" % (filename,
```
single quotes please, and please try to use `git commit -a --amend`  -- these extra commits are not necessary.

dhruvsomani replied:
Oops! Sorry, I made this change and amended this in the previous commit.

myint commented on scspell/__init__.py:
```diff
@@ -556,13 +581,32 @@ def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+        errors = list(errors.items())
+        errors.sort(key=lambda x: x[1])
+
+        print('Set of mispelled tokens:')
```
The rest of the code seems to print to `stderr`, so could you do the same here as well (`print(..., file=sys.stderr`)?

myint commented on scspell/__init__.py:
```diff
@@ -556,13 +581,32 @@ def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+
+        print('Set of mispelled tokens:')
+        for row in errors:
+            print("%s:[%s]: %s" % (filename,
```
Same comment as above. Print to `stderr`.

myint reviewed (changes requested):
Thanks!

- [ ] Similar to my comment in #31, add a test case to `test.cram`.

dhruvsomani committed:
Added option to list only unique misspellings.

Added the `--unique-list` command line argument which enables to only see the unique errors in a report only
manner. Thus, recurring not matching tokens will be lumped together as one. Also, shows the corresponding line
numbers where the misspelling occurs.

The `list_unique` parameter was added to `spell_check()`, `spell_check_file()` and `spell_check_token()`
functions.

Closes https://github.com/myint/scspell/issues/24

dhruvsomani commented:
@myint @jayvdb I changed the output file to `sys.stderr` and also reverted the wrong changes I had made for the backslash issue. So, I think this can be merged now.

myint commented on scspell/__init__.py:
```diff
@@ -556,13 +566,32 @@ def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+        errors = list(errors.items())
+        errors.sort(key=lambda x: x[1])
+
+        print('Set of mispelled tokens:', file=sys.stderr)
```
"mispelled" is a misspelling. 😄

myint commented on scspell/__init__.py:
```diff
@@ -499,12 +505,13 @@ def spell_check_token(
 
 
-def spell_check_file(filename, dicts, ignores, report_only, c_escapes):
+def spell_check_file(filename, dicts, ignores, report_only, c_escapes, list_unique):
```
Put `list_unique` on the next line to make this fit within 80 characters.

myint commented on scspell/__init__.py:
```diff
@@ -487,9 +487,15 @@ def spell_check_token(
-            if report_only:
+            if list_unique:
+                return ((match_desc.get_string(),
+                        match_desc.get_ofs() + len(token)),
```
This is missing one additional space of indentation.

myint reviewed (changes requested):
A few more requested after I tested this with `make check`.

dhruvsomani commented:
@myint Done. 😄

myint commented:
The indentation is still a bit off. But I"m going approve as I suspect this Google thing will be already over the next time I get a chance to log in.

```
$ pycodestyle scspell/__init__.py
scspell/__init__.py:495:29: E127 continuation line over-indented for visual indent
scspell/__init__.py:515:25: E127 continuation line over-indented for visual indent
scspell/__init__.py:594:37: E127 continuation line over-indented for visual indent
```
